### PR TITLE
Fix backward multi round hashing

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Hey @lucidrains,

Awesome work on reformer-pytorch. Your code helped me a lot to integrate Reformer into the hugging face library (here: https://huggingface.co/transformers/model_doc/reformer.html), so hopefully this PR helps you as well :-) 

I am quite confident that your code as it is now contains a bug so that the gradients are wrong for multi-round hashing (`n_hashes > 1`). 

In the customized backward of `UnsortLogits`, the gradients are currently re-sorted over all hash dimensions, but they should be re-sorted for each hash dimension individually.

 In the hugging face reformer code, I added gradient integration tests comparing the gradients to the original trax implementation. Here the code in current master:
https://github.com/huggingface/transformers/blob/1af58c07064d8f4580909527a8f18de226b226ee/tests/test_modeling_reformer.py#L931

or this branch for the full testing suite I had when developing the code:
https://github.com/huggingface/transformers/blob/branch_to_save_trax_integration_tests/tests/test_modeling_reformer.py

This PR should fix this error when calculating the gradients for `n_hashes > 1`.

I used this small script to test my changes:

```python
#!/usr/bin/env python3
import torch
from torch import randint

from reformer_pytorch import ReformerLM
from reformer_pytorch.generative_tools import TrainingWrapper

model = ReformerLM(
    num_tokens= 200,
    dim = 48,
    depth = 2,
    max_seq_len = 128,
    lsh_dropout = 0.1,
    causal = True,
    n_hashes=4
)

# 0 is used for padding and no loss to be calculated on it
model = TrainingWrapper(model, ignore_index = 0, pad_value = 0)

# the wrapper can handle evenly packed sequences
x_train = randint(0, 200, (128,)).unsqueeze(0)

# when training, set return_loss equal to True
model.train()

loss = model(x_train, return_loss = True)
loss.backward()
```

Maybe @nkitaev might be able to comment here as well :-)

### CODE REVIEW

**Comment by lucidrains** [comment]
@patrickvonplaten thanks for reviewing the code and glad to see it has made it's way into huggingface!

It's been a while since I wrote that code, but I transcribed it almost verbatim from https://github.com/google/trax/blob/master/trax/layers/research/efficient_attention.py#L1265-L1266

Let me review this next week and get back to you!

**Comment by patrickvonplaten** [comment]
@lucidrains - I reviewed the code again and I think your implementation as it was before is correct. Sorry to have bothered you!

The implementation proposed in this PR is correct as well, but obviously not as clean as the one you had before.
I added a PR to make the code cleaner in the `transformers` implementation as well according to your implementation and ran all the gradient tests against the original trax implementation => looks good, so the code as you had it before is definitely the way to go :-)

**Comment by lucidrains** [comment]
@patrickvonplaten no problem! i'm just glad there's someone joining me. there is very few people in the entire world working on this!
